hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java

```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
       return ctx;
     }
 
+    /***
+     * a thread for parallel MD5 computing to increase performance when loading
+     */
+    private static class DigestThread extends Thread {
+      volatile private IOException ioe = null;
+      volatile private MD5Hash digest = null;
+      private File file;
+
+      public DigestThread(File inFile) {
+        file = inFile;
+      }
+
+      public MD5Hash getDigest() {
+        return digest;
+      }
+
+      public IOException getException() {
+        return ioe;
+      }
+
+      @Override
+      public void run() {
+        try {
+          digest = MD5FileUtils.computeMd5ForFile(file);
+        } catch (IOException e) {
+          ioe = e;
+        } catch (Throwable t) {
+          ioe = new IOException(t);
+        }
+      }
+    }
+
     void load(File file) throws IOException {
       long start = Time.monotonicNow();
-      imgDigest = MD5FileUtils.computeMd5ForFile(file);
+      DigestThread dt = new DigestThread(file);
```

Comment:
Should this be a daemon?

Reply:
Actually how expensive are these two steps? (digest calculation and load)

Reply:
Test the patch against a fsimage of a 70PB  cluster (200million files and 300million blocks), the image loading time be reduced from 1210 seconds to 1105 seconds.

Reply:
That's fair, can you extend the description of the JIRA saying it can reduce up to 10% of the time or similar?